docs/proposals/resource-quota.md

```diff
@@ -0,0 +1,404 @@
+}
+
+// A field selector operator is the set of operators that can be used in a selector requirement.
+type FieldSelectorOperator string
+
+const (
+  FieldSelectorOpIn           FieldSelectorOperator = "In"
+  FieldSelectorOpNotIn        FieldSelectorOperator = "NotIn"
+  FieldSelectorOpExists       FieldSelectorOperator = "Exists"
+  FieldSelectorOpDoesNotExist FieldSelectorOperator = "DoesNotExist"
+)
+```
+
+### ResourceQuota
+
+The `ResourceQuotaSpec` will be modified to support the optional
+selectors to filter the `ResourceQuota` document.
+
+```
+// ResourceQuotaSpec defines the desired hard limits to enforce for Quota
+type ResourceQuotaSpec struct {
+  // Hard is the set of desired hard limits for each named resource
+  Hard ResourceList `json:"hard,omitempty"`
+  // LabelSelector is a label query that must match over each resource that is tracked by quota.
+  // If empty, the quota matches every resource.
+  // More info: http://releases.k8s.io/HEAD/docs/user-guide/labels.md#label-selectors
+  LabelSelector *LabelSelector `json:"labelSelector,omitempty"`
+  // FieldSelector is a field query that must match over the resource tracked by quota.
+  // If empty, the quota matches every resource.
+  FieldSelector *FieldSelector `json:"fieldSelector,omitempty"`
+}
+```
+
+The selectors are immutable post-creation.
+
+API definition of `ReplenishmentPolicy` to support replenishment
+via a rate are deferred pending further prototyping to understand
+what additional fields may or may not be needed to track when
+quota was last replenished.
+
+### Pod
+
+The administrator is able to filter terminating vs non-terminating workloads
+by defining a field selector of `spec.restartPolicy`.
+
+The administrator is not able to filter pods by quality of service.  To
+facilitate this ability, the `Pod` should expose a new field on it's `PodSpec`
+that corresponds to its `QualityOfService` that is set by the server at
+creation and used by `ResourceQuota`.  The administrator would then
+be able to filter quota by tracking `spec.qualityOfService` to quota
+`BestEffort` pods differently than `Burstable` and `Guaranteed`.
+
+## Implementation Phases and Anticipated Issues
+
+Phase 1:
+
+* Support ability for cluster-admin to quota based on workload duration
+ * Add `spec.restartAlways` as a field selector for `Pod`
+ * Ensure API server support for set-based and equality based matching of field selector
+ * Add `FieldSelector` to `v1` API
+ * Add `FieldSelector` to `ResourceQuotaSpec` to enable tracking on restart policy
+ * Add logic to admission control and quota controller to enable support
+
+* Improve latency of quota controller observing freed resources for non-pod types
+by watching for more resource types
+
+* Identify and solve issues when running large numbers of namespaces and quotas
+ * Target 20k namespaces with 60k quota documents as baseline scenario
+ * Minimize contention on locking of quota documents at admission
+
+Phase 2:
```

Comment:
Agreed.  I noted it here since by default I imagine pod storage strategy would apply its value prior to persistence.  Given the fact that ResourceQuota admission control runs prior to the storage strategy, the value would be calculated there as well ( since we know requests and limits at this point).  Not running quota, value would be set in strategy.